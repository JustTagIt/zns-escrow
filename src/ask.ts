import {createInterface} from 'readline';

const readlineInterface = createInterface(process.stdin, process.stdout);

export default async function ask(question = 'Continue?') {
  let answered;
  do {
    answered = await new Promise(resolve => {
      readlineInterface.question(question + ' [Y/n] ', answer => {
        if (/^(?:[Y|y](?:es)?)?$/.test(answer)) {
          resolve(true);
        } else if (/^(?:[N|n]o?)?$/.test(answer)) {
          console.log('Stopping...');
          process.exit(1);
        } else {
          resolve(false);
        }
      });
    });
  } while (!answered);
}
