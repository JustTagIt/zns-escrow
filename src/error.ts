export default function error(message, code = 1) {
  console.error('zns-escrow:', message);
  process.exit(code);
}
