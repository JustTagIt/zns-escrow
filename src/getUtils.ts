import {toBech32Address} from '@zilliqa-js/crypto';
import {bytes} from '@zilliqa-js/util';
import {Zilliqa} from '@zilliqa-js/zilliqa';

let hasLoggedBefore = false;

export default function getUtils({privateKey, verbose, url, chainId}) {
  const zilliqa = new Zilliqa(url);
  const address = zilliqa.wallet.addByPrivateKey(privateKey);
  zilliqa.wallet.setDefault(address);

  const version = bytes.pack(chainId, 1);

  if (verbose && !hasLoggedBefore) {
    hasLoggedBefore = true;

    console.info('Using url:', url);
    console.info('Using version:', version);
    console.info('Using address:', address);
    console.info('Using bech32 address:', toBech32Address(address));
  }

  return {address, version, zilliqa};
}
