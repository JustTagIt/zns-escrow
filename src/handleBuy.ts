import buy from './actions/buy';
import {buyPreflight} from './preflight';

export default function handleBuy(argv) {
  const {
    escrow,
    privateKey,
    domain,
    verbose,
    url,
    chainId,
    registry,
    node,
  } = buyPreflight(argv);

  return buy({
    escrow,
    privateKey,
    domain,
    verbose,
    url,
    chainId,
    registry,
    node,
  });
}
