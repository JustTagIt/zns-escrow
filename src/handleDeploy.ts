import {deployPreflight} from './preflight';

export default function handleDeploy(argv) {
  const {
    privateKey,
    domain,
    verbose,
    url,
    chainId,
    registry,
    node,
    buyer,
    price,
    seller,
  } = deployPreflight(argv);

  console.log('deploy', {
    privateKey,
    domain,
    verbose,
    url,
    chainId,
    registry,
    node,
    buyer,
    price,
    seller,
  });
}
