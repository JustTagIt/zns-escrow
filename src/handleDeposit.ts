import {depositPreflight} from './preflight';

export default function handleDeposit(argv) {
  const {
    escrow,
    privateKey,
    domain,
    verbose,
    url,
    chainId,
    registry,
  } = depositPreflight(argv);

  console.log('deposit', {
    escrow,
    privateKey,
    domain,
    verbose,
    url,
    chainId,
    registry,
  });
}
