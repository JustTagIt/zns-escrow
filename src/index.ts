import {
  getAddressFromPrivateKey,
  normaliseAddress,
  toBech32Address,
} from '@zilliqa-js/crypto';
import chalk from 'chalk';
import namehash from 'namicorn/lib/zns/namehash';
import * as yargs from 'yargs';

function error(message, code = 1) {
  console.error('zns-escrow:', message);
  process.exit(code);
}

function preflight<T extends {} = {}>({
  domain,
  url,
  chainId,
  registry,
  privateKey,
  verbose = false,
  ...rest
}): {
  domain: string;
  node: string;
  url: string;
  chainId: number;
  registry: string;
  privateKey: string;
  verbose: boolean;
} & T {
  if (!/^[a-f\d]{64}|[A-F\d]{64}$/.test(privateKey)) {
    error('bad private key');
  }

  let node;
  try {
    node = namehash(domain);
  } catch (error) {
    error('bad domain');
  }

  let registryAddress;
  if (registry) {
    try {
      registryAddress = normaliseAddress(registryAddress);
    } catch (error) {
      error('bad registry');
    }
  }

  switch (chainId) {
    // mainnet
    case 1: {
      return {
        verbose,
        domain,
        node,
        url: url || 'https://api.zilliqa.com/',
        chainId,
        registry:
          registryAddress ||
          normaliseAddress(
            toBech32Address('0x9611c53be6d1b32058b2747bdececed7e1216793'),
          ),
        privateKey,
        ...rest,
      } as any;
    }
    // testnet
    case 111: {
      if (!registry) {
        error('must supply registry using testnet');
      }

      return {
        verbose,
        domain,
        node,
        url: url || 'https://dev-api.zilliqa.com/',
        chainId,
        registry,
        privateKey,
        ...rest,
      } as any;
    }
    // kaya
    case 333: {
      error('cannot use kaya');
    }
    default: {
      if (!registry) {
        error('must supply registry using private chain');
      }

      if (!url) {
        error('must supply url using private chain');
      }

      return {
        verbose,
        domain,
        node,
        url,
        chainId,
        registry,
        privateKey,
        ...rest,
      } as any;
    }
  }
}

function deployPreflight(argv) {
  const {
    privateKey,
    domain,
    verbose,
    url,
    chainId,
    registry,
    node,
    buyer,
    price,
    qa = false,
    seller = getAddressFromPrivateKey(privateKey),
  } = preflight<{buyer: string; price: number; qa?: boolean; seller?: string}>(
    argv,
  );

  let buyerAddress: string;
  try {
    buyerAddress = normaliseAddress(buyer);
  } catch (error) {
    error('bad buyer');
  }

  let sellerAddress: string;
  try {
    sellerAddress = normaliseAddress(seller);
  } catch (error) {
    error('bad seller');
  }

  if (!Number.isInteger(price) || price < 1) {
    error('bad price, must be positive integer');
  }

  return {
    privateKey,
    domain,
    verbose,
    url,
    chainId,
    registry,
    node,
    buyer: buyerAddress,
    price: qa ? price : price * 10 ** 12,
    seller: sellerAddress,
  };
}

function depositPreflight(argv) {
  const {
    privateKey,
    domain,
    verbose,
    url,
    chainId,
    registry,
    escrow,
  } = preflight<{escrow: string}>(argv);

  let escrowAddress: string;
  try {
    escrowAddress = normaliseAddress(escrow);
  } catch (error) {
    error('bad escrow');
  }

  return {
    escrow: escrowAddress,
    privateKey,
    domain,
    verbose,
    url,
    chainId,
    registry,
  };
}

function handleSell(argv) {
  const {
    privateKey,
    domain,
    verbose,
    url,
    chainId,
    registry,
    node,
    buyer,
    price,
    seller,
  } = deployPreflight(argv);

  console.log('sell', {
    privateKey,
    domain,
    verbose,
    url,
    chainId,
    registry,
    node,
    buyer,
    price,
    seller,
  });
}

function handleDeploy(argv) {
  const {
    privateKey,
    domain,
    verbose,
    url,
    chainId,
    registry,
    node,
    buyer,
    price,
    seller,
  } = deployPreflight(argv);

  console.log('deploy', {
    privateKey,
    domain,
    verbose,
    url,
    chainId,
    registry,
    node,
    buyer,
    price,
    seller,
  });
}

function handleDeposit(argv) {
  const {
    escrow,
    privateKey,
    domain,
    verbose,
    url,
    chainId,
    registry,
  } = depositPreflight(argv);

  console.log('deposit', {
    escrow,
    privateKey,
    domain,
    verbose,
    url,
    chainId,
    registry,
  });
}

function handleBuy(argv) {
  const {
    escrow,
    privateKey,
    domain,
    verbose,
    url,
    chainId,
    registry,
  } = depositPreflight(argv);

  console.log('buy', {
    escrow,
    privateKey,
    domain,
    verbose,
    url,
    chainId,
    registry,
  });
}

yargs
  .help()
  .demandCommand()
  .option('verbose', {
    alias: 'v',
    type: 'boolean',
    description: 'Run with verbose logging',
  })
  .option('private-key', {
    alias: 'k',
    type: 'string',
    description: 'Use private key for transactions',
  })
  .option('url', {
    alias: 'u',
    type: 'string',
    description: 'Use Url for Zilliqa api',
  })
  .option('chain-id', {
    type: 'number',
    description: 'Use custom chain id for Zilliqa',
    default: 1,
  })
  .option('registry', {
    type: 'string',
    description: 'Use ZNS Registry contract address',
  })
  .demandOption('private-key')
  .epilog('Made with ' + chalk.red('â™¥') + ' by Unstoppable Domains')
  .command(
    'sell <domain>',
    'Deploy escrow contract',
    yargs =>
      yargs
        .positional('domain', {
          type: 'string',
          describe: 'domain to sell',
        })
        .option('buyer', {
          type: 'string',
          description: 'Buyer address',
        })
        .option('price', {
          type: 'number',
          description: 'Price to sell at',
        })
        .option('qa', {
          type: 'boolean',
          description: 'Use Qa instead of Zillings for price',
        })
        .option('seller', {
          type: 'string',
          description: 'Seller address',
        })
        .demandOption(['buyer', 'price']),
    handleSell,
  )
  .command(
    'deploy <domain>',
    'Deploy escrow contract',
    yargs =>
      yargs
        .positional('domain', {describe: 'domain to sell'})
        .option('buyer', {
          type: 'string',
          description: 'Buyer address',
        })
        .option('price', {
          type: 'number',
          description: 'Price to sell at',
        })
        .option('qa', {
          type: 'boolean',
          description: 'Use Qa instead of Zillings for price',
        })
        .option('seller', {
          type: 'string',
          description: 'Seller address',
        })
        .demandOption(['buyer', 'price']),
    handleDeploy,
  )
  .command(
    'deposit <domain>',
    'Deposit domain on contract',
    yargs =>
      yargs
        .positional('domain', {describe: 'domain to sell'})
        .option('escrow', {
          type: 'string',
          description: 'Escrow contract address',
        })
        .demandOption('escrow'),
    handleDeposit,
  )
  .command(
    'buy <domain>',
    'Purchase domain from contract',
    yargs =>
      yargs
        .positional('domain', {describe: 'domain to sell'})
        .option('escrow', {
          type: 'string',
          description: 'Escrow contract address',
        })
        .demandOption('escrow'),
    handleBuy,
  ).argv;
