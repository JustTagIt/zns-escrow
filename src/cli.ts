import chalk from 'chalk';
import * as yargs from 'yargs';
import handleBuy from './handleBuy';
import handleDeploy from './handleDeploy';
import handleDeposit from './handleDeposit';
import handleSell from './handleSell';

yargs
  .help()
  .demandCommand()
  .option('verbose', {
    alias: 'v',
    type: 'boolean',
    description: 'Run with verbose logging',
  })
  .option('private-key', {
    alias: 'k',
    type: 'string',
    description: 'Use private key for transactions',
  })
  .option('url', {
    alias: 'u',
    type: 'string',
    description: 'Use Url for Zilliqa api',
  })
  .option('chain-id', {
    type: 'number',
    description: 'Use custom chain id for Zilliqa',
    default: 1,
  })
  .option('registry', {
    type: 'string',
    description: 'Use ZNS Registry contract address',
  })
  .demandOption('private-key')
  .epilog('Made with ' + chalk.red('â™¥') + ' by Unstoppable Domains')
  .command(
    // ['sell <domain>', '$0 <domain>'],
    'sell <domain>',
    'Deploy escrow contract and deposit name onto contract',
    yargs =>
      yargs
        .positional('domain', {
          type: 'string',
          describe: 'domain to sell',
        })
        .option('buyer', {
          type: 'string',
          description: 'Buyer address',
        })
        .option('price', {
          type: 'number',
          description: 'Price to sell at',
        })
        .option('qa', {
          type: 'boolean',
          description: 'Use Qa instead of Zillings for price',
        })
        .option('seller', {
          type: 'string',
          description: 'Seller address',
        })
        .demandOption(['buyer', 'price']),
    handleSell,
  )
  .command(
    'deploy <domain>',
    'Deploy escrow contract',
    yargs =>
      yargs
        .positional('domain', {describe: 'domain to sell'})
        .option('buyer', {
          type: 'string',
          description: 'Buyer address',
        })
        .option('price', {
          type: 'number',
          description: 'Price to sell at',
        })
        .option('qa', {
          type: 'boolean',
          description: 'Use Qa instead of Zillings for price',
        })
        .option('seller', {
          type: 'string',
          description: 'Seller address',
        })
        .demandOption(['buyer', 'price']),
    handleDeploy,
  )
  .command(
    'deposit <domain>',
    'Deposit domain on contract',
    yargs =>
      yargs
        .positional('domain', {describe: 'domain to sell'})
        .option('escrow', {
          type: 'string',
          description: 'Escrow contract address',
        })
        .demandOption('escrow'),
    handleDeposit,
  )
  .command(
    'buy <domain>',
    'Purchase domain from contract',
    yargs =>
      yargs
        .positional('domain', {describe: 'domain to sell'})
        .option('escrow', {
          type: 'string',
          description: 'Escrow contract address',
        })
        .demandOption('escrow'),
    handleBuy,
  ).argv;
