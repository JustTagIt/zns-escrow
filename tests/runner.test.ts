import {Value} from '@zilliqa-js/contract/src/types';
import {execSync} from 'child_process';
import {existsSync, mkdirSync, readFileSync, writeFileSync} from 'fs';
import {join} from 'path';

const node =
  '0x0000000000000000000000000000000000000000000000000000000000000000';
const seller = '0x1111111111111111111111111111111111111111';
const buyer = '0x2222222222222222222222222222222222222222';
const registry = '0x3333333333333333333333333333333333333333';
const thisAddress = '0x4444444444444444444444444444444444444444';
const address = '0x5555555555555555555555555555555555555555';
const price = '10';

const init = [
  {
    vname: '_scilla_version',
    type: 'Uint32',
    value: '0',
  },
  {
    vname: 'registry',
    type: 'ByStr20',
    value: registry,
  },
  {
    vname: 'seller',
    type: 'ByStr20',
    value: seller,
  },
  {
    vname: 'buyer',
    type: 'ByStr20',
    value: buyer,
  },
  {
    vname: 'escrowedNode',
    type: 'ByStr32',
    value: node,
  },
  {
    vname: 'price',
    type: 'Uint128',
    value: price,
  },
  {
    vname: '_this_address',
    type: 'ByStr20',
    value: thisAddress,
  },
  {
    vname: '_creation_block',
    type: 'BNum',
    value: '0',
  },
];

function createState({
  _balance = 0 as number | string,
  sold = false,
  deposit = false,
} = {}) {
  return [
    {
      vname: '_balance',
      type: 'Uint128',
      value: _balance.toString(),
    },
    {
      vname: 'id',
      type: 'String',
      value: 'udc:escrow:1',
    },
    {
      vname: 'sold',
      type: 'Bool',
      value: {
        constructor: sold ? 'True' : 'False',
        argtypes: [],
        arguments: [],
      },
    },
    {
      vname: 'deposit',
      type: 'Bool',
      value: {
        constructor: deposit ? 'True' : 'False',
        argtypes: [],
        arguments: [],
      },
    },
  ];
}

function scillaRun({
  init: initArr = init,
  state,
  message,
  blockchain = [{vname: 'BLOCKNUMBER', type: 'BNum', value: '1'}],
  gaslimit = 10000,
}: {
  init?: Value[] | never[];
  state?: Value[] | never[];
  message?: {
    _tag: string;
    _amount: string;
    _sender: string;
    params: Value[] | never[];
  };
  blockchain?: Value[];
  gaslimit?: number | string;
} = {}) {
  if (!existsSync('/tmp/zns-escrow')) {
    mkdirSync('/tmp/zns-escrow');
  }

  writeFileSync('/tmp/zns-escrow/init.json', JSON.stringify(initArr));
  if (state) {
    writeFileSync('/tmp/zns-escrow/state.json', JSON.stringify(state));
  }
  if (message) {
    writeFileSync('/tmp/zns-escrow/message.json', JSON.stringify(message));
  }
  writeFileSync('/tmp/zns-escrow/blockchain.json', JSON.stringify(blockchain));

  const result = execSync(`scilla-runner \
      -i ${join(__dirname, '../contracts/escrow.scilla')} \
      -init /tmp/zns-escrow/init.json \
      ${state ? '-istate /tmp/zns-escrow/state.json \\' : '\\'}
      ${message ? '-imessage /tmp/zns-escrow/message.json \\' : '\\'}
      -iblockchain /tmp/zns-escrow/blockchain.json \
      -o /tmp/zns-escrow/output.json \
      -gaslimit ${gaslimit} \
      -libdir ${join(__dirname, '../contracts/stdlib')}
  `);

  return JSON.parse(readFileSync('/tmp/zns-escrow/output.json', 'utf8'));
}

it('should deploy', () => {
  const output = scillaRun();

  expect(output.states).toEqual(createState());
});

it('should deposit', () => {
  const message = {
    _tag: 'deposit',
    _amount: price,
    _sender: buyer,
    params: [],
  };

  const output = scillaRun({
    message,
    state: createState(),
  });

  expect(output._accepted).toBe('true');
  expect(output.states.find(v => v.vname === 'deposit').value).toEqual({
    argtypes: [],
    arguments: [],
    constructor: 'True',
  });
  expect(output.states.find(v => v.vname === '_balance').value).toBe(price);
  expect(output.message).toBeNull();

  expect(output).toMatchSnapshot('deposit_success');
});

it('should swap', () => {
  const message = {
    _tag: 'swap',
    _amount: '0',
    _sender: address,
    params: [],
  };

  const output = scillaRun({
    message,
    state: createState({
      deposit: true,
    }),
  });

  expect(output.message._recipient).toBe(registry);
  expect(output.message._tag).toBe('transfer');
  expect(output.message.params.find(v => v.vname === 'owner').value).toBe(
    buyer,
  );

  expect(output).toMatchSnapshot('swap_success');
});

it('should fail to deposit with incorrect price', () => {
  const message = {
    _tag: 'deposit',
    _amount: '1',
    _sender: buyer,
    params: [],
  };

  const output = scillaRun({
    message,
    state: createState(),
  });

  expect(output._accepted).toBe('false');
  expect(output.states.find(v => v.vname === '_balance').value).toBe('0');
  expect(output.states.find(v => v.vname === 'deposit').value).toEqual({
    argtypes: [],
    arguments: [],
    constructor: 'False',
  });

  expect(output.message).toBeNull();

  expect(output).toMatchSnapshot('deposit_failure');
});

it('should payout seller on transfer success', () => {
  const message = {
    _tag: 'onTransferSuccess',
    _amount: '0',
    _sender: registry,
    params: [
      {vname: 'owner', type: 'ByStr20', value: buyer},
      {vname: 'node', type: 'ByStr32', value: node},
    ],
  };

  const output = scillaRun({
    message,
    state: createState({_balance: price}),
  });

  expect(output._accepted).toBe('false');
  expect(output.states.find(v => v.vname === 'sold').value).toEqual({
    argtypes: [],
    arguments: [],
    constructor: 'True',
  });
  expect(output.message._recipient).toBe(seller);
  expect(output.message._amount).toBe(price);

  expect(output).toMatchSnapshot('onTransferSuccess_success');
});

it('should emit error event on transfer success if not registry', () => {
  const message = {
    _tag: 'onTransferSuccess',
    _amount: '0',
    _sender: address,
    params: [
      {vname: 'owner', type: 'ByStr20', value: buyer},
      {vname: 'node', type: 'ByStr32', value: node},
    ],
  };

  const output = scillaRun({
    message,
    state: createState(),
  });

  expect(output._accepted).toBe('false');
  expect(output.message).toBeNull();
  expect(output.events.find(e => e._eventname === 'Error')).toBeTruthy();

  expect(output).toMatchSnapshot('onTransferSuccess_failure');
});

it('should refund on transfer fail', () => {
  const message = {
    _tag: 'onTransferFailure',
    _amount: '0',
    _sender: registry,
    params: [
      {vname: 'owner', type: 'ByStr20', value: buyer},
      {vname: 'node', type: 'ByStr32', value: node},
    ],
  };

  const output = scillaRun({
    message,
    state: createState({_balance: price}),
  });

  expect(output._accepted).toBe('false');
  expect(output.message._recipient).toBe(buyer);
  expect(output.message._amount).toBe(price);

  expect(output).toMatchSnapshot('onTransferFailure_success');
});

it('should emit error event on transfer fail if not registry', () => {
  const message = {
    _tag: 'onTransferFailure',
    _amount: '0',
    _sender: address,
    params: [
      {vname: 'owner', type: 'ByStr20', value: buyer},
      {vname: 'node', type: 'ByStr32', value: node},
    ],
  };

  const output = scillaRun({
    message,
    state: createState(),
  });

  expect(output._accepted).toBe('false');
  expect(output.message).toBeNull();
  expect(output.events.find(e => e._eventname === 'Error')).toBeTruthy();

  expect(output).toMatchSnapshot('onTransferFailure_success');
});
