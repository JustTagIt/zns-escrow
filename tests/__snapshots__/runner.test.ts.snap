// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should deposit: deposit_success 1`] = `
Object {
  "_accepted": "true",
  "events": Array [
    Object {
      "_eventname": "Deposit",
      "params": Array [],
    },
  ],
  "gas_remaining": "9753",
  "message": null,
  "scilla_major_version": "0",
  "states": Array [
    Object {
      "type": "Uint128",
      "value": "10",
      "vname": "_balance",
    },
    Object {
      "type": "Bool",
      "value": Object {
        "argtypes": Array [],
        "arguments": Array [],
        "constructor": "True",
      },
      "vname": "deposit",
    },
    Object {
      "type": "String",
      "value": "udc:escrow:1",
      "vname": "id",
    },
    Object {
      "type": "Bool",
      "value": Object {
        "argtypes": Array [],
        "arguments": Array [],
        "constructor": "False",
      },
      "vname": "sold",
    },
  ],
}
`;

exports[`should emit error event on transfer fail if not registry: onTransferFailure_success 1`] = `
Object {
  "_accepted": "false",
  "events": Array [
    Object {
      "_eventname": "Error",
      "params": Array [],
    },
  ],
  "gas_remaining": "9593",
  "message": null,
  "scilla_major_version": "0",
  "states": Array [
    Object {
      "type": "Uint128",
      "value": "0",
      "vname": "_balance",
    },
    Object {
      "type": "String",
      "value": "udc:escrow:1",
      "vname": "id",
    },
    Object {
      "type": "Bool",
      "value": Object {
        "argtypes": Array [],
        "arguments": Array [],
        "constructor": "False",
      },
      "vname": "sold",
    },
    Object {
      "type": "Bool",
      "value": Object {
        "argtypes": Array [],
        "arguments": Array [],
        "constructor": "False",
      },
      "vname": "deposit",
    },
  ],
}
`;

exports[`should emit error event on transfer success if not registry: onTransferSuccess_failure 1`] = `
Object {
  "_accepted": "false",
  "events": Array [
    Object {
      "_eventname": "Error",
      "params": Array [],
    },
  ],
  "gas_remaining": "9593",
  "message": null,
  "scilla_major_version": "0",
  "states": Array [
    Object {
      "type": "Uint128",
      "value": "0",
      "vname": "_balance",
    },
    Object {
      "type": "String",
      "value": "udc:escrow:1",
      "vname": "id",
    },
    Object {
      "type": "Bool",
      "value": Object {
        "argtypes": Array [],
        "arguments": Array [],
        "constructor": "False",
      },
      "vname": "sold",
    },
    Object {
      "type": "Bool",
      "value": Object {
        "argtypes": Array [],
        "arguments": Array [],
        "constructor": "False",
      },
      "vname": "deposit",
    },
  ],
}
`;

exports[`should fail to deposit with incorrect price: deposit_failure 1`] = `
Object {
  "_accepted": "false",
  "events": Array [
    Object {
      "_eventname": "Error",
      "params": Array [],
    },
  ],
  "gas_remaining": "9758",
  "message": null,
  "scilla_major_version": "0",
  "states": Array [
    Object {
      "type": "Uint128",
      "value": "0",
      "vname": "_balance",
    },
    Object {
      "type": "String",
      "value": "udc:escrow:1",
      "vname": "id",
    },
    Object {
      "type": "Bool",
      "value": Object {
        "argtypes": Array [],
        "arguments": Array [],
        "constructor": "False",
      },
      "vname": "sold",
    },
    Object {
      "type": "Bool",
      "value": Object {
        "argtypes": Array [],
        "arguments": Array [],
        "constructor": "False",
      },
      "vname": "deposit",
    },
  ],
}
`;

exports[`should payout seller on transfer success: onTransferSuccess_success 1`] = `
Object {
  "_accepted": "false",
  "events": Array [
    Object {
      "_eventname": "Success",
      "params": Array [],
    },
  ],
  "gas_remaining": "9455",
  "message": Object {
    "_amount": "10",
    "_recipient": "0x1111111111111111111111111111111111111111",
    "_tag": "",
    "params": Array [],
  },
  "scilla_major_version": "0",
  "states": Array [
    Object {
      "type": "Uint128",
      "value": "0",
      "vname": "_balance",
    },
    Object {
      "type": "Bool",
      "value": Object {
        "argtypes": Array [],
        "arguments": Array [],
        "constructor": "True",
      },
      "vname": "sold",
    },
    Object {
      "type": "String",
      "value": "udc:escrow:1",
      "vname": "id",
    },
    Object {
      "type": "Bool",
      "value": Object {
        "argtypes": Array [],
        "arguments": Array [],
        "constructor": "False",
      },
      "vname": "deposit",
    },
  ],
}
`;

exports[`should refund on transfer fail: onTransferFailure_success 1`] = `
Object {
  "_accepted": "false",
  "events": Array [],
  "gas_remaining": "9495",
  "message": Object {
    "_amount": "10",
    "_recipient": "0x2222222222222222222222222222222222222222",
    "_tag": "",
    "params": Array [],
  },
  "scilla_major_version": "0",
  "states": Array [
    Object {
      "type": "Uint128",
      "value": "0",
      "vname": "_balance",
    },
    Object {
      "type": "Bool",
      "value": Object {
        "argtypes": Array [],
        "arguments": Array [],
        "constructor": "True",
      },
      "vname": "deposit",
    },
    Object {
      "type": "String",
      "value": "udc:escrow:1",
      "vname": "id",
    },
    Object {
      "type": "Bool",
      "value": Object {
        "argtypes": Array [],
        "arguments": Array [],
        "constructor": "False",
      },
      "vname": "sold",
    },
  ],
}
`;

exports[`should swap: swap_success 1`] = `
Object {
  "_accepted": "false",
  "events": Array [],
  "gas_remaining": "9637",
  "message": Object {
    "_amount": "0",
    "_recipient": "0x3333333333333333333333333333333333333333",
    "_tag": "transfer",
    "params": Array [
      Object {
        "type": "ByStr32",
        "value": "0x0000000000000000000000000000000000000000000000000000000000000000",
        "vname": "node",
      },
      Object {
        "type": "ByStr20",
        "value": "0x2222222222222222222222222222222222222222",
        "vname": "owner",
      },
    ],
  },
  "scilla_major_version": "0",
  "states": Array [
    Object {
      "type": "Uint128",
      "value": "0",
      "vname": "_balance",
    },
    Object {
      "type": "String",
      "value": "udc:escrow:1",
      "vname": "id",
    },
    Object {
      "type": "Bool",
      "value": Object {
        "argtypes": Array [],
        "arguments": Array [],
        "constructor": "False",
      },
      "vname": "sold",
    },
    Object {
      "type": "Bool",
      "value": Object {
        "argtypes": Array [],
        "arguments": Array [],
        "constructor": "True",
      },
      "vname": "deposit",
    },
  ],
}
`;
